### PR TITLE
Blob unit and integration testing

### PR DESCRIPTION
**Pull request by pomeroy3**
Split apart blob unit tests from integration tests and update the testing section in the README.

### CODE REVIEW

File: gr-azure-software-radio/README.md [hunk 1]
```diff
@@ -43,45 +43,55 @@ NOTE: If you have installed the Azure CLI with the default apt package on Ubuntu
...
     ```
-    python -m pytest qa_*
+    python -m pytest integration_*
     ```
     
     or
 
     ```
-    python3 -m unittest qa_*
+    python3 -m unittest integration_*
```
Comment (bquantump) [inline gr-azure-software-radio/README.md:94]: Anyway to run with ctest? Just curios .

Comment (pomeroy3) [reply gr-azure-software-radio/README.md:94]: I specifically went against having it work with ctest. I don't think we want the integration tests running for 'make test' and I didn't have a clean way to have ctest run certain tests conditionally. 

If there's value in getting it working we can put something in the backlog.

**Comment by markosly** [comment]
Looks good to me, my only question is if we will be enforcing flake8 on integration/qa files. If so I think there might be a few changes needed.

**Review by bquantump — CHANGES_REQUESTED** [review]

File: gr-azure-software-radio/python/blob_source.py [hunk 2]
```diff
@@ -131,6 +140,11 @@ def work(self, input_items, output_items):
             int: Number of items generated in this work call
         """
 
+        # connect to the blob service the first time we run
```
[the remark this answers is not in the document]
Comment (pomeroy3) [reply gr-azure-software-radio/python/blob_source.py:143]: This made the code a little more unit-test friendly by removing things that actually talk to Azure out of the constructor and pushing them into the first work call.

File: gr-azure-software-radio/python/integration_blob_source.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+        self.tb.connect(op, dst)
+        # set up fg
+        self.tb.run()
+        # check data
+
+        result_data = dst.data()
+
+        # check data
+        self.assertTrue((src_data == result_data).all())
```
[the remark this answers is not in the document]
Comment (pomeroy3) [reply gr-azure-software-radio/python/integration_blob_source.py:82]: looks like that doesn't like arrays?

```
>       self.assertEqual(src_data, result_data)
E       ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

File: gr-azure-software-radio/python/qa_blob_source.py [hunk 4]
```diff
@@ -147,8 +105,37 @@ def test_chunk_residue_merge(self):
...
+        src_data = np.arange(0, num_samples, 1, dtype=np.complex64)
+
+        dst = blocks.vector_sink_c()
+
+        # prevent setup_blob_iterator from making Azure API calls
+        with patch.object(blob_source, 'setup_blob_iterator', spec=iter) as mock_iter:
+            # add in a list of chunks we want to pretend the blob API gave us
+            mock_iter.return_value = iter([src_data.tobytes()])
+            
+            # set up a blob sink
```
[the remark this answers is not in the document]
Comment (pomeroy3) [reply gr-azure-software-radio/python/qa_blob_source.py]: fixed

File: gr-azure-software-radio/python/qa_blob_source.py [hunk 5]
```diff
@@ -147,8 +105,37 @@ def test_chunk_residue_merge(self):
...
+    def test_end_to_end_run(self):
+        '''
+        Test the block properly starts up, reads data from the blob data queue, and cleanly
+        shuts down
+        '''
 
+        blob_name = 'test-blob.npy'
+        num_samples = 500
+
+        # set up a vector source with known complex data
```
[the remark this answers is not in the document]
Comment (pomeroy3) [reply gr-azure-software-radio/python/qa_blob_source.py]: fixed

File: gr-azure-software-radio/python/qa_blob_source.py [hunk 6]
```diff
@@ -10,68 +10,41 @@
...
+        instance = blob_source(authentication_method="connection_string",
+                               connection_str=self.blob_connection_string,
+                               container_name=self.test_blob_container_name,
+                               blob_name='test-instance',
+                               queue_size=4)
 
-        result_data = dst.data()
+        # really only checking that the init didn't throw an exception above, but adding the check
+        # below to keep flake8 happy
+        self.assertTrue(instance is not None)
```
[the remark this answers is not in the document]
Comment (pomeroy3) [reply gr-azure-software-radio/python/qa_blob_source.py]: fixed

File: gr-azure-software-radio/python/integration_blob_source.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (pomeroy3) [reply gr-azure-software-radio/python/integration_blob_source.py:79]: Only plays nice if you also convert all arrays to lists. No biggie either way

File: gr-azure-software-radio/python/integration_blob_sink.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        # connect to the test blob container and download the file
+        # compare the file we downloaded against the samples in the vector source
+        blob_client = self.blob_service_client.get_blob_client(
+            container=self.test_blob_container_name,
+            blob=blob_name)
+
+        result_data = np.frombuffer(blob_client.download_blob().readall(), dtype=np.complex64)
+
+        # check data
+        self.assertTrue((src_data == result_data).all())
```
[the remark this answers is not in the document]
Comment (pomeroy3) [reply gr-azure-software-radio/python/integration_blob_sink.py]: fixed

**Review by bquantump — APPROVED** [review]

**Review by markosly — APPROVED** [review]

**Pull request merged**